### DISCUSSION TITLE
Smart Meter Interface - EasyMeter Q3A

### DISCUSSION THREAD

**Discussion by Pewibe** [arendst/Tasmota#12597] [Q&A] [answered]
This call for help is dedicated to Smart Meter Interface, especially model EasyMeter Q3A. On the MSB-Interface (on top of the meter) I receive a telegram once per second like this:
```
02:32:36.501 : 77 01 0b 09 01 45 53 59 11 03 b0 6b 09 07 01 00 62 0a ff ff 72 62 01 65 00 23 34 5f 79 
02:32:36.544 : 77 07 01 00 00 00 09 ff 01 01 01 01 0b 09 01 45 53 59 11 03 b0 6b 09 01 
02:32:36.572 : 77 07 01 00 01 08 00 ff 64 00 00 80 01 62 1e 52 fc 59 00 00 00 00 9e f6 4e ba 01 (Arbeit Summe)
02:32:36.597 : 77 07 01 00 10 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 9c 0b 01 (Leistung Summe)
02:32:36.621 : 77 07 01 00 24 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 53 ae 01 (Leistung Phase L1) 
02:32:36.646 : 77 07 01 00 38 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 42 fd 01 (Leistung Phase L2)
02:32:36.672 : 77 07 01 00 4c 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 05 5e 01 (Leistung Phase L3)
02:32:36.519 : 77 07 81 81 c7 82 03 ff 01 01 01 01 04 45 53 59 01 
02:32:36.738 : 77 07 81 81 c7 82 05 ff 01 01 01 01 83 02 0d ca d7 5a ff b9 54 80 92 72 7d 01 37 d6 b6 0c 08 8d d3 c1 d3 9f d9 f4 bb a4
02:32:36.478 : 77 07 81 81 c7 f0 06 ff 01 01 01 01 04 31 00 0e 01 01 01 63 88 e5 00 76 0b 45 53 59 41 6b 09 00 69 9a 11 62 00 62 00 72 

```
Hint 1: Text in paranthesis is my comment.
Hint 2: The 3-byte-gap in the lines (Leistung ...) I added for better alignment to the line above.

I use this script:
```
>D
>B
=>sensor53 r
>M 1
+1,3,s,0,9600,ENO
1,77078181c78203ff@#,Herstellerkennung,,Manu,0 
1,77070100100700ff@1000,Summe_Leistung,kW,SumPower,0
1,77070100240700ff@1000,Ph1_Leistung,kW,SumPower1,0
1,77070100380700ff@1000,Ph2_Leistung,kW,SumPower2,0
1,770701004c0700ff@1000,Ph3_Leistung,kW,SumPower3,0
1,77070100010800ff@1000,Summe_Verbrauch,kWh,SumEnergy,0
#

```
On the main web page I see:
```
Sonoff BasicR3
H.StromZaehler.106
ENO Herstellerkennung:	"455359"
ENO Summe_Leistung:	0 kW
ENO Ph1_Leistung:	0 kW
ENO Ph2_Leistung:	0 kW
ENO Ph3_Leistung:	0 kW
ENO Summe_Verbrauch:	282 kWh
. . .
Tasmota 9.5.0.1 by Theo Arends

```
**1st question:**
The missing 4 power values are because of a smart meter bug or because of a buggy tasmota module or because of a wrong script? Any idea?

**2nd question:**
Line '02:32:36.519' contains the so Hersteller-Kennung / manufacturer ID (3 byte 45 53 59). Is it possible to see "ESY" instead of "455359" and how do I have to modificate the corresponding line of the script?

Be patient with my missing knowledge and thanks for your help in advance.
To inform you personally: @nagyrobi @MichiM2

**Comment by MichiM2** [comment, marked as answer]
Hi,

I use the "Info-Schnittstelle INFO-DSS" (the leftmost IR-LED on the front). The MSB-DSS is sealed on my Q3A :-(
I use this script with Tasmota 9.2.0 :


 >D
 >B
=>sensor53 r
 >M 1
+1,3,s,0,9600,SML
***@***.***,Verbrauch_Tarif_1,kWh,Total_Tarif1,0
***@***.***,Verbrauch_Tarif_2,kWh,Total_Tarif2,0
***@***.***,Verbrauch_Summe,kWh,Total_Summe,0
#


and it works as expected.



I have no Hexdump available.


Michi



Am 08.07.21 um 18:43 schrieb Pewibe:

**Reply by Pewibe** [reply]
@MichiM2 What you wrote is known by me (INFO interface), but my question is dedicated to the parsing of the 4 lines 'Leistung ...'. 
By the way: you have a hex dump available if you enter the command 'sensor53 d1'.

**Reply by Pewibe** [reply]
Additional hint: The MSB-DSS at my meter was not usable like yours. But the 'optical cover' I interpretate as 'removable sticker' and not as 'seal'. Live is so easy! :-)

**Comment by Pewibe** [comment]
@gemu2015 Dear Gerhard
As author of the modul code I like to address you personally to read (and hopefully answer) my questions above. Thanks!

**Reply by gemu2015** [reply]
these values are actually in Watts, so you should try with 3 decimals:

1,77070100240700ff@1000,Ph1_Leistung,kW,SumPower1,3
1,77070100380700ff@1000,Ph2_Leistung,kW,SumPower2,3
1,770701004c0700ff@1000,Ph3_Leistung,kW,SumPower3,3

or in Watts
1,77070100240700ff@1,Ph1_Leistung,W,SumPower1,0
1,77070100380700ff@1,Ph2_Leistung,W,SumPower2,0
1,770701004c0700ff@1,Ph3_Leistung,W,SumPower3,0

**Reply by Pewibe** [reply]
Thanks, your answer was helpful and I see the power values now. 

Above there was a 2nd question - can you answer it:
Line '02:32:36.519' contains the Hersteller-Kennung / manufacturer ID (3 byte 45 53 59). Is it possible to see "ESY" instead of "455359" and how do I have to modificate the corresponding line of the script?

In the meantime more questions came up - are you the right address to answer?
a) I find it helpful to have a LED blinking if a block of data telegrams from smart meter is received (usually every second). Sometimes I'm missing telegrams if the magnetic head is not placed correct at smart meter or other reasons. Another idea would be to enable by command 'sensor53 d2' a line saying '15:13:57.980 XXX: data telegram(s) received'. This might be better than a screen full of hex code.
b) Are there any existing functions or future plans for building blended values (e.g. average power in one minute, one hour, ...)?
c) Are there any existing functions or future plans for sending commands towards the smart meter to control functions?

**Reply by gemu2015** [reply]
1. Hersteller Kennung is not supported as string, only service id
2. there is a Variable smlv[x] that is set when a value is decoded. (and reset with smlv[0]) with a little script you can determine if your variables are still updated.
3. all kinds of results including graphs are supported with the scripting language.
4. you may also send commands to the meter.

have a look at the tasmota docs for sml and scripting. there are various examples.

**Reply by Pewibe** [reply]
Answer 1, 3, 4 are understood. Concerning 2 I didn't discover smlv in source code and %smlv[1]% didn't work in the script. Can you check it and maybe write some script lines to show how it works? Thanks!

**Reply by gemu2015** [reply]
#define USE_SML_SCRIPT_CMD

is needed for this cmd.


this code snipped checks every 10 seconds if a line was decoded:

\>S
; every 10 seconds
if upsecs%10==0
then
; check if decode line 1 was decoded
if smlv[1]==0
then
print no value decoded within 10 seconds
endif
; reset all flags
smlv[0]
endif

**Reply by Pewibe** [reply]
@gemu2015  I compiled with #define USE_SML_SCRIPT_CMD and was wondering that 
`print %smlv[1]%`
didn't work because you wrote it's a variable. I extended the script section to your proposal
```
>S
if upsecs%5==0
then
if smlv[1]==0 
then 
print smlv[1] = 0
else 
print smlv[1] = 1
endif
print %smlv[1]%
smlv[0]
endif
```
Independant of the infrared head is attached to the smart meter or not, I always see the same lines
```
12:46:23.342 script compressed to 427 bytes = 61 %
12:46:23.344 Script: nv=0, tv=0, vns=0, ram=16
12:46:23.346 Script: performs "sensor53 r"
12:46:23.354 SNS: Hardware Serial
12:46:23.359 MQT: stat/H.StromZaehler/RESULT = {"Time":"2021-07-17T12:46:23","SML":{"CMD":"restart"}}
12:46:23.617 smlv[1] = 1
12:46:23.618 ???
12:46:28.630 smlv[1] = 1
12:46:28.632 ???
12:46:33.602 smlv[1] = 1
12:46:33.604 ???
```
What goes wrong?

**Reply by gemu2015** [reply]
did you use a very recent dev version ?
it should not display ???

**Reply by Pewibe** [reply]
@gemu2015  
No, I used the source of 9.5.0 as base to enable SML. by 'it should not display ???' you mean that smlv[x] is a legal variable and line 'print %smlv[1]%' should work?

Because it is hard to debug by writing Q&A, maybe you can explain where to find this smlv[x] construct in source (I serched for it but didn't catch it). And how slmv[0] works without any assignment like 'smlv[0]=x'? Why didn't you design 'smlv[x]=0' to reset the decoding flag? I guess having more than one smart meter, it is not the best to reset the flag for all meters at the same time.

Over all, in my opinion it is not the best to check the validation of decoded smart meter data by extra scripts. Such a status should be displayed on web page and send by MQTT in conjunction with the values. I hope I don't bother you with my thoughts.

**Reply by Pewibe** [reply]
@gemu2015 Are you absent or did I wrote something wrong? Maybe you can answer the  problem that I do not find 'smlv' in the code to make my own thoughts - thanks!

**Reply by gemu2015** [reply]
what i tried to say was that if you get ??? you are using an outdated version. use the most recent dev version.

**Reply by Pewibe** [reply]
@gemu2015 As I wrote 13 days ago, I used the source package of 9.5.0 which is the actual version still today. Any other hint?

**Comment by gemu2015** [comment]
this is NOT the latest version. you need at least 9.5.02
you must use the latest DEV version!

https://github.com/arendst/Tasmota/blob/development/CHANGELOG.md

**Comment by VinniF** [comment]
> This call for help is dedicated to Smart Meter Interface, especially model EasyMeter Q3A. On the MSB-Interface (on top of the meter) I receive a telegram once per second like this:
> 
> ```
> 02:32:36.501 : 77 01 0b 09 01 45 53 59 11 03 b0 6b 09 07 01 00 62 0a ff ff 72 62 01 65 00 23 34 5f 79 
> 02:32:36.544 : 77 07 01 00 00 00 09 ff 01 01 01 01 0b 09 01 45 53 59 11 03 b0 6b 09 01 
> 02:32:36.572 : 77 07 01 00 01 08 00 ff 64 00 00 80 01 62 1e 52 fc 59 00 00 00 00 9e f6 4e ba 01 (Arbeit Summe)
> 02:32:36.597 : 77 07 01 00 10 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 9c 0b 01 (Leistung Summe)
> 02:32:36.621 : 77 07 01 00 24 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 53 ae 01 (Leistung Phase L1) 
> 02:32:36.646 : 77 07 01 00 38 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 42 fd 01 (Leistung Phase L2)
> 02:32:36.672 : 77 07 01 00 4c 07 00 ff 01          01 62 1b 52 fe 59 00 00 00 00 00 00 05 5e 01 (Leistung Phase L3)
> 02:32:36.519 : 77 07 81 81 c7 82 03 ff 01 01 01 01 04 45 53 59 01 
> 02:32:36.738 : 77 07 81 81 c7 82 05 ff 01 01 01 01 83 02 0d ca d7 5a ff b9 54 80 92 72 7d 01 37 d6 b6 0c 08 8d d3 c1 d3 9f d9 f4 bb a4
> 02:32:36.478 : 77 07 81 81 c7 f0 06 ff 01 01 01 01 04 31 00 0e 01 01 01 63 88 e5 00 76 0b 45 53 59 41 6b 09 00 69 9a 11 62 00 62 00 72 
> ```
> 
> Hint 1: Text in paranthesis is my comment. Hint 2: The 3-byte-gap in the lines (Leistung ...) I added for better alignment to the line above.
> 
> I use this script:
> 
> ```
> >D
> >B
> =>sensor53 r
> >M 1
> +1,3,s,0,9600,ENO
> 1,77078181c78203ff@#,Herstellerkennung,,Manu,0 
> 1,77070100100700ff@1000,Summe_Leistung,kW,SumPower,0
> 1,77070100240700ff@1000,Ph1_Leistung,kW,SumPower1,0
> 1,77070100380700ff@1000,Ph2_Leistung,kW,SumPower2,0
> 1,770701004c0700ff@1000,Ph3_Leistung,kW,SumPower3,0
> 1,77070100010800ff@1000,Summe_Verbrauch,kWh,SumEnergy,0
> #
> ```
> 
> On the main web page I see:
> 
> ```
> Sonoff BasicR3
> H.StromZaehler.106
> ENO Herstellerkennung:	"455359"
> ENO Summe_Leistung:	0 kW
> ENO Ph1_Leistung:	0 kW
> ENO Ph2_Leistung:	0 kW
> ENO Ph3_Leistung:	0 kW
> ENO Summe_Verbrauch:	282 kWh
> . . .
> Tasmota 9.5.0.1 by Theo Arends
> ```
> 
> **1st question:** The missing 4 power values are because of a smart meter bug or because of a buggy tasmota module or because of a wrong script? Any idea?
> 
> **2nd question:** Line '02:32:36.519' contains the so Hersteller-Kennung / manufacturer ID (3 byte 45 53 59). Is it possible to see "ESY" instead of "455359" and how do I have to modificate the corresponding line of the script?
> 
> Be patient with my missing knowledge and thanks for your help in advance. To inform you personally: @nagyrobi @MichiM2

Hi, did you ever solve this problem? My smartmeter does has an IR I/O but it doesn't have the option to activate additional info (despite entering the PIN). So I removed the sticker and now I want to try it via the MSB interface. Can you help there what that protocol looks like?

**Comment by gemu2015** [comment]
if you post your dump at this url:
https://tasmota-sml-parser-dev.azurewebsites.net/decode

you get this descriptor:

M 1
+1,3,s,0,9600,
1,77070100000009FF@1,Geräteeinzelidentifikation,Unbekannte Einheit,geräteeinzelidentifikation,0
1,77070100010800FF@1,Zählerstand Bezug Total,Wh,zählerstand_bezug_total,4
1,77070100100700FF@1,aktuelle Wirkleistung,W,aktuelle_wirkleistung,2
1,77070100240700FF@1,Summenwirkleistung L1,W,summenwirkleistung_l1,2
1,77070100380700FF@1,Summenwirkleistung L2,W,summenwirkleistung_l2,2
1,770701004C0700FF@1,Summenwirkleistung L3,W,summenwirkleistung_l3,2

**Comment by VinniF** [comment]
Where do I get the dump from? I am using the stromleser with tasmota. The command sensor53 d1 gives  {"Command":"Error","Input":"SENSOR53 d1"}

Edit: Just start running a random script first, then the command will be successful
